dashboard/streamlit_graphs.py

```diff
@@ -0,0 +1,175 @@
+"""Script to make the bar chart of top sales by genre."""
+from os import environ
+import streamlit as st
+import psycopg2
+from datetime import datetime
+from psycopg2 import extensions
+import pandas as pd
+import altair as alt
+
+from dotenv import load_dotenv
+import logging
+
+# Adjust Streamlit logger to suppress errors or specific logs
+logger = logging.getLogger('streamlit')
+logger.setLevel(logging.CRITICAL)
+
+# Your app code
+st.write("Errors from Streamlit won't appear here.")
+
+
+load_dotenv()
+
+
+def get_connection() -> extensions.connection | None:
+    """
+    Tries to connect to the RDS database.
+    """
+    try:
+        connection = psycopg2.connect(
+            host=environ["DB_HOST"],
+            port=environ["DB_PORT"],
+            user=environ["DB_USER"],
+            password=environ["DB_PASSWORD"],
+            database=environ["DB_NAME"]
+        )
+        return connection
+    except psycopg2.OperationalError:
+        return None
+
+
+def fetch_sales_within_date_range(connection, start_date, end_date) -> pd.DataFrame | None:
+    """
+    Fetches sales data aggregated by hour for a given date range.
+    """
+    query = """
+    SELECT
+        DATE_TRUNC('hour', sale_date) AS sale_hour,
+        COUNT(sale_id) AS total_sales
+    FROM sale
+    WHERE sale_date BETWEEN %s AND %s
+    GROUP BY sale_hour
+    ORDER BY sale_hour;
+    """
+    try:
+        sale_data = pd.read_sql(query, connection, params=[
+                                start_date, end_date])
+
+        if not sale_data.empty:
+            return sale_data
+        else:
+            return None
+    except Exception as e:
+        logging.error("Error fetching sales data: %s", e)
+        return None
+
+
+def get_sales_data(connection: extensions.connection) -> pd.DataFrame:
+    """
+    Returns the total sales of the top
+    5 performing genres as a dataframe.
+    """
+    query = """
+    SELECT
+        g.genre_name,
+        SUM(s.sale_price) AS total_sales
+    FROM
+        sale s
+    JOIN
+        release r ON s.release_id = r.release_id
+    JOIN
+        release_genre rg ON r.release_id = rg.release_id
+    JOIN
+        genre g ON rg.genre_id = g.genre_id
+    GROUP BY
+        g.genre_name
+    ORDER BY
+        total_sales DESC
+    LIMIT 5;
+    """
+
+    return pd.read_sql_query(query, connection)
+
+
+def create_genre_sales_chart(connection: extensions.connection) -> alt.Chart | None:
+    """
+    Generates a bar chart for genre sales data.
+    """
+
+    sales_data = get_sales_data(connection)
+
+    if sales_data.empty:
+        st.warning("No data available to display.")
+        return None
+
+    sales_data = sales_data.sort_values("total_sales", ascending=False)
+    sales_data["rank"] = range(1, len(sales_data) + 1)
+
+    custom_colors = ["#2596be", "#51abcb",
```

Comment:
Maybe these could be made into magic numbers so we can easily change the global colour scheme?

Reply:
It's a little hard to do this, as I need as many custom colours as there are items in the x-axis, the moment we change the amount, to something like 10, we'd need to add more colours.